# bcr-lineage
This repository includes scripts for preprocessing MiXCR output files, defining Ig clonal groups and building lineage trees.

### Preprocess MiXCR output
*R/parse_mixcr.R* contains functions for reading MiXCR clonotype files and extracting mutations (either from alignment or from mutationsDetailed* if available).

The following code reads all the clonotype files from *mixcr_out_folder*, reads germline V and J sequences from *germline_v_path* and *germline_j_path* respectively and returns a list with parsed clonotype, substitution and indel tables (saved as .Rds files in *out_path*):
```R
source('R/parse_mixcr.R')
mxr <- read_mixcr(mixcr_out_folder = 'example_files/data/mixcr_mut_detailed/', out_path = 'example_files/out', germline_v_path = 'example_files/germline/germline_ighv.csv', germline_j_path = 'example_files/germline/germline_ighj.csv', ncores = 2)
```
Germline sequences are needed when mutationsDetailed field is unavailable and amino acid replacements are defined by comparison to the germline sequence.

### Define clonal groups
*R/clusterize.R* takes clonotype table as input, runs MIR and defines cluster (cluster_id added as a new column to clonotype table). Example:
```R
source('R/clusterize.R')
define_clusters(clone_table_path = 'clones.Rds', output_path = 'clusters', mir_path = '../',
ncores = 4)
```

parameter       | description
----------------|-------------
clone_table_path | path to *clones.Rds* generated by read_mixcr()
output_path | folder to write output cluster table and to store temporary files for MIR
mir_path | path to MIR binary file
ncores | number of cores to use
save_mir_files | whether to keep temporary files for MIR (TRUE/FALSE)
merge_by_root | whether to merge clusters additionaly by aligning root sequences (TRUE/FALSE)





